README.rst
==========
PyGrowCube
==========


.. image:: https://img.shields.io/pypi/v/pygrowcube.svg
        :target: https://pypi.python.org/pypi/pygrowcube

.. image:: https://img.shields.io/travis/robotedward/pygrowcube.svg
        :target: https://travis-ci.com/robotedward/pygrowcube

.. image:: https://readthedocs.org/projects/pygrowcube/badge/?version=latest
        :target: https://pygrowcube.readthedocs.io/en/latest/?version=latest
        :alt: Documentation Status


.. image:: https://pyup.io/repos/github/robotedward/pygrowcube/shield.svg
     :target: https://pyup.io/repos/github/robotedward/pygrowcube/
     :alt: Updates



Python module for interacting with Elecrow GrowCube smart plant watering system

* Free software: MIT license
* Documentation: https://pygrowcube.readthedocs.io.


Features
--------

* Basic CLI and module to read current GrowCube version, temperature, humidity and status of each conencted moisture sensor

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2023-09-02)
------------------

* First release on PyPI.

0.2.0 (2023-09-12)
------------------

* Switched to asyncio for networking

0.2.1 (2023-09-26)
------------------

* Handle low water error


tests/__init__.py
"""Unit test package for pygrowcube."""


tests/test_pygrowcube.py
#!/usr/bin/env python

"""Tests for `pygrowcube` package."""

import pytest

from click.testing import CliRunner

from pygrowcube import pygrowcube
from pygrowcube import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    assert 'pygrowcube.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


pygrowcube/cli.py
"""Console script for pygrowcube."""
import asyncio
import sys
import click
from pygrowcube.pygrowcube import get_status
import logging


@click.command()
@click.argument("ip_address")
@click.option(
    "--timeout",
    "-t",
    default=15,
    show_default=True,
    help="Maximum time to wait for readings in seconds. GrowCube typically sends readings within 10s.",
)
@click.option(
    "--connect",
    "-c",
    is_flag=True,
    default=False,
    help="Validate connection to GrowCube without waiting for sensor readings",
)
@click.option(
    "--verbose", "-v", is_flag=True, default=False, help="Send log output to stdout"
)
@click.option("--debug", "-d", is_flag=True, default=False, help="Enable debug logging")
@click.option(
    "--log", "-l", is_flag=True, default=False, help="Send log output to a file"
)
@click.option(
    "--logfilename",
    "-f",
    default="pygrowcube.log",
    help="Specify a filename to use for log output",
)
def main(timeout, ip_address, connect, verbose, debug, log, logfilename):
    """Get the status and sensor readings of the GrowCube device at IP_ADDRESS"""
    loglevel = logging.INFO if verbose else logging.ERROR
    loglevel = logging.DEBUG if debug else loglevel
    if log:
        logging.basicConfig(
            filename=logfilename,
            level=loglevel,
            format="%(asctime)s %(levelname)s:%(message)s",
        )
    else:
        logging.basicConfig(
            stream=sys.stdout,
            level=loglevel,
            format="%(asctime)s %(levelname)s:%(message)s",
        )

    status = asyncio.run(get_status(ip_address, timeout, not connect))

    click.echo(str(status))
    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


pygrowcube/pygrowcube.py
"""Main module."""
from .message import Message
from .message import MessageType
from .messageclient import MessageClient
from .timeouthelper import TimeoutHelper
import logging

PORT = 8800
STATUS_TIMEOUT = (
    15  # wait max 15 seconds - sensors send a refresh every 10s when connected
)

logger = logging.getLogger(__name__)


class Status:
    def __init__(
        self,
        temperature=0,
        humidity=0,
        moistures=[0, 0, 0, 0],
        sensor_warnings=[0, 0, 0, 0],
        outlet_locks=[0, 0, 0, 0],
        version="",
        id="",
        host="",
        has_water=True,
        connect_only=False,
    ):
        self.temperature = temperature
        self.humidity = humidity
        self.moistures = moistures
        self.sensor_warnings = sensor_warnings
        self.outlet_locks = outlet_locks
        self.version = version
        self.refreshed_sensors = [False, False, False, False]
        self.id = id
        self.host = host
        self.connect_only = connect_only
        self.has_water = has_water

    def __str__(self) -> str:
        s = f"GrowCube {self.id} ({self.host}). Software version: {self.version}\n"
        if not self.connect_only:
            s += f"Temperature: {self.temperature}, Humidity: {self.humidity}\n"
            if not self.has_water:
                s += "Warning: Not enough water. Refill and press the unlock button on GrowCube device.\n"
            for i in range(4):
                s += f" - Sensor {i}: "
                if self.sensor_warnings[i]:
                    s += f"DISCONNECTED "
                else:
                    if self.outlet_locks[i]:
                        s += "OUTLET LOCKED "
                    if not self.refreshed_sensors[i]:
                        s += "NO READING "
                    else:
                        s += f"{self.moistures[i]}"
                s += "\n"
            if not self.is_refresh_complete:
                s += f"Warning: GrowCube did not send latest status for all sensors before we stopped waiting\n"
        return s.rstrip()

    @property
    def is_refresh_complete(self):
        return all(self.refreshed_sensors)

    def handle_sensor_disconnected(self, message: Message):
        if not message.message_content.isdigit():
            raise ValueError(
                f"{message.readable_message_type}: Expecting message content to be a channel number. Message:"
                + message.get_message()
            )
        channel = int(message.message_content)
        if not channel < 4:
            raise ValueError(
                f"{message.readable_message_type}: Expecting channel number to be less than 4. Message:"
                + message.get_message()
            )
        self.sensor_warnings[channel] = True

    def handle_outlet_locked(self, message: Message):
        if not message.message_content.isdigit():
            raise ValueError(
                f"{message.readable_message_type}: Expecting message content to be a channel number. Message:"
                + message.get_message()
            )
        channel = int(message.message_content)
        if not channel < 4:
            raise ValueError(
                f"{message.readable_message_type}: Expecting channel number to be less than 4. Message:"
                + message.get_message()
            )
        self.outlet_locks[channel] = True

    def handle_sensor_reading(self, message: Message):
        fields = message.get_fields()
        assert len(fields) == 4, (
            "Expecting message content to have 4 fields: " + message.get_message()
        )
        assert all(field.isdigit() for field in fields), (
            "Expecting message content fields to be all digits" + message.get_message()
        )
        channel, reading, humidity, temperature = fields
        channel = int(channel)
        assert channel < 4, "Channel number out of range: " + message.get_message()
        self.humidity = int(humidity)
        self.temperature = int(temperature)
        self.moistures[channel] = int(reading)
        self.refreshed_sensors[channel] = True
        logger.debug(f"Received reading for sensor {channel}: {reading}.")

    def handle_start_reading(self, message: Message):
        self.refreshed_sensors = [False, False, False, False]
        self.moistures = [0, 0, 0, 0]
        if not (
            (message.message_content == "0@0") or (message.message_content == "1@1")
        ):
            raise ValueError(
                f"{message.readable_message_type}: Received content other than 0@0 or 1@1. Message: {message.get_message()}"
            )
        if message.message_content == "1@1":
            self.has_water = False

    def handle_growcube_version(self, message: Message):
        version, id = message.message_content.split("@")
        self.version = version
        self.id = id

    def handle_OK(self, message: Message):
        if message.message_content != "1":
            logger.warn(
                f"{message.readable_message_type} received with unexpected content: {message.message_content}"
            )
        return

    def default_handler(self, message: Message):
        logger.warn(
            f"{message.readable_message_type} default handler called. Full message: {message.get_message()}"
        )
        return

    status_handlers = {
        MessageType.VERSION: handle_growcube_version,
        MessageType.START_READINGS: handle_start_reading,
        MessageType.SENSOR_READING: handle_sensor_reading,
        MessageType.SENSOR_DISCONNECTED: handle_sensor_disconnected,
        MessageType.OUTLET_LOCKED: handle_outlet_locked,
        MessageType.OK: handle_OK,
    }

    def handle_message(self, message: Message):
        logger.debug(
            f"RECEIVED {message.readable_message_type}: {message.get_message()}"
        )
        if message.message_type in self.status_handlers:
            handler = self.status_handlers.get(message.message_type)
            handler(self, message)
        else:
            self.default_handler(message)


async def get_status(
    growcube_address: str,
    timeout_in_seconds: float = STATUS_TIMEOUT,
    wait_for_sensor_readings: bool = True,
) -> Status:
    logger.info(
        f"Getting status of GrowCube at {growcube_address}:{PORT}. Timeout {timeout_in_seconds}. Wait for readings: {wait_for_sensor_readings}."
    )
    client = MessageClient(growcube_address, PORT)
    status = Status(host=growcube_address, connect_only=not wait_for_sensor_readings)
    timeout = TimeoutHelper(timeout_in_seconds)
    try:
        await client.connect()
        request = Message(
            message_type=MessageType.REQUEST_HELLO,
            message_content=Message.format_datetime_for_growcube(),
        )
        await client.send_message(request, timeout)
        response = await client.receive_message(timeout)
        if response.message_type != MessageType.VERSION:
            logger.error(
                f"Did not receive version number as expected. Response: {str(response)}"
            )
        else:
            status.handle_message(response)
            if wait_for_sensor_readings:
                request = Message(
                    message_type=MessageType.REQUEST_READINGS, message_content="2"
                )
                await client.send_message(request, timeout)
                while not status.is_refresh_complete:
                    response = await client.receive_message(timeout)
                    if isinstance(response, Message):
                        status.handle_message(response)
                    else:
                        logger.warn(
                            f"Response is not a recognisable message: {str(response)}"
                        )
                    if timeout.timed_out:
                        logger.warn(
                            "Did not get a complete refresh of all sensors within time out"
                        )
                        break
                    else:
                        logger.debug(
                            f"Looping. Elapsed:{timeout.elapsed}, Timeout: {timeout_in_seconds}"
                        )
            return status
    finally:
        await client.close()


pygrowcube/messageclient.py
import asyncio
import logging
from .timeouthelper import TimeoutHelper
from .message import Message
from .message import MessageType
from time import perf_counter

logger = logging.getLogger(__name__)
TIMEOUT = 5


class MessageClient:
    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.reader = None
        self.writer = None

    async def connect(self):
        try:
            logger.debug("Connecting to: %s %s ", self.host, self.port)
            self.reader, self.writer = await asyncio.wait_for(
                asyncio.open_connection(self.host, self.port), timeout=TIMEOUT
            )
        except asyncio.TimeoutError:
            logger.exception(
                f"Connection timed out connecting to: {self.host} {self.port}"
            )
            raise
        except Exception as e:
            logger.exception(
                f"Connection error: {e}. Connecting to: {self.host} {self.port}"
            )
            raise

    async def close(self):
        if self.writer:
            self.writer.close()
            await asyncio.wait_for(self.writer.wait_closed(), timeout=TIMEOUT)

    async def send_message(
        self, message: Message, timeout: TimeoutHelper = TimeoutHelper(TIMEOUT)
    ):
        if not self.writer:
            raise ValueError(
                "Socket connection is not established. Call connect() first."
            )
        try:
            message_string = message.get_message()
            logger.info(
                f"SENDING {message.readable_message_type}: {message.message_content}. {message_string}"
            )
            self.writer.write(message_string.encode())
            await asyncio.wait_for(self.writer.drain(), timeout=timeout.remaining)
        except asyncio.TimeoutError:
            logger.exception("Network operation timed out.")
            raise
        except Exception as e:
            logger.exception(f"Error sending message: {e}")
            raise

    async def read_until_delimiter(
        self, timeout: TimeoutHelper = TimeoutHelper(TIMEOUT), delimiter=b"#"
    ) -> str:
        """Read a field from growcube until we get to a delimiter character.
        We have to do this a character at a time to strip the zeros sent in between messages.
        """
        data = b""
        try:
            while True:
                if timeout.timed_out:
                    logger.warn(
                        "Timed out waiting for data. Timeout=%s, elapsed=%s. Received: %s",
                        timeout.timeout,
                        timeout.elapsed,
                        data.decode(),
                    )
                    raise asyncio.TimeoutError()
                chunk = await asyncio.wait_for(
                    self.reader.read(1), timeout=timeout.remaining
                )
                if chunk == delimiter:
                    break
                if chunk != (b"\x00"):
                    data += chunk
                if len(data) == 6:
                    # GrowCube sends ele5XX responses without a trailing #
                    s = data[:4].decode()
                    if s == "ele5":
                        break
        except asyncio.TimeoutError:
            logger.warn(
                "Timed out waiting for data. Timeout=%s, elapsed=%s. Received: %s",
                timeout.timeout,
                timeout.elapsed,
                data.decode(),
            )
            raise
        except Exception as e:
            logger.exception(f"Error reading data {e}")
            raise
        return data.decode()

    async def receive_message(
        self, timeout: TimeoutHelper = TimeoutHelper(TIMEOUT)
    ) -> Message:
        if not self.reader:
            raise ValueError(
                "Socket connection is not established. Call connect() first."
            )
        try:
            # Receive message type and content length values
            message_type_string = await self.read_until_delimiter(timeout)
            if not message_type_string.startswith("ele"):
                raise ValueError(
                    f"Did not recognise a GrowCube message type - does not begin with 'ele': {message_type}"
                )
            if message_type_string[3] == "a":
                message_type = int(message_type_string[4:])
            else:
                message_type = int(message_type_string[3:])
            message = Message(message_type=message_type)
            if not message.content_expected_for_message_type:
                return message

            length = await self.read_until_delimiter(timeout)
            if not length.isdigit():
                raise ValueError(
                    f"Did not receive an integer content length as expected: {length}"
                )

            # Receive the message content based on the content length
            message_content_response = (await asyncio.wait_for(
                self.reader.read(int(length) + 1), timeout=timeout.remaining
            )).decode()
            message_content = message_content_response[:-1]
            end = message_content_response[-1]
            if not end == "#":
                logger.warn(
                    "Unexpected content at end of message. Expecting #, got " + end
                )
            message.content_length = length
            message.message_content = message_content
            logger.info(f"RECEIVED {message.readable_message_type}: {message_content}")
            return message
        except Exception as e:
            logger.error(f"Error receiving message: {e}")
            return None


# Example usage:
# if __name__ == "__main__":
#     host = "example.com"  # Replace with the actual remote host address
#     port = 12345  # Replace with the actual port number

#     client = MessageClient(host, port)

#     try:
#         client.connect()

#         # Receive and parse a message from the remote host
#         received_data = client.receive_message()
#         if received_data:
#             received_message = Message(received_data)
#             print("Received Message:")
#             print(received_message)
#     finally:
#         client.close()


pygrowcube/message.py
from datetime import datetime
from enum import IntEnum
import logging


class MessageType(IntEnum):
    # Messages without conten:
    REQUEST_READY_FOR_COMMAND = (506,)
    READY_FOR_COMMAND = 550

    # Client command messages:
    REQUEST_READINGS = (
        43  # TBC: Sent by the client close to start of session elea43#1#2#
    )
    REQUEST_HELLO = 44  # Sent with client datetime at start of session elea44#19#2023@09@05@11@53@30#
    REQUEST_DELETE = 45  # Delete the settings for a channel - always seems to be sent as a pair with 46 - elea45#1#0#
    REQUEST_DELETE_CONFIRM = 46  # Second part of the delete request - don't know why but client alwasy sends this immediately after #45
    REQUEST_WATER_CONTROL = 47  # Turn water on or off for a channel
    REQUEST_SENSOR_HISTORY = 48  # Request the moisture history for a channel
    REQUEST_CHANNEL_SETTINGS = 49  # Settings for a channel
    # Smart watering: channel@3@min%@max% elea49#9#2@3@10@50#
    # Smart watering outside of sunlight hours: elea49#9#2@2@10@50#
    # Regular watering: channel@1@how_often@how_long elea49#9#2@1@30s@6#

    # GrowCube response messages:
    OK = 20  # suspect this is OK - always comes back with a #1
    SENSOR_READING = 21  # Moisture reading for an individual sensor
    SENSOR_HISTORY_ENTRY = (
        22  # Moisture readings by hour for a specific channel and date
    )
    WATERING_HISTORY_ENTRY = (
        23  # Invdividual watering event date and time for a specific channel
    )
    VERSION = 24  # Sent on connection
    WATER_ON = 26  # Water turned on for a channel
    WATER_OFF = 27  # Water turned off for a channel
    SENSOR_DISCONNECTED = (
        30  # Sent on connection when a sensor is disconnected elea30#1#2#
    )
    START_READINGS = 33  # Always sent before sensor readings with content:
    # 0@0 if normal.
    # 1@0 if device is locked due to low water
    OUTLET_LOCKED = 34  # The outlet for the specified channel is locked elea34#1#1#
    UNKNOWN = 0


class Message:
    def __init__(
        self,
        message_string=None,
        message_type: MessageType = MessageType.UNKNOWN,
        message_content=None,
    ):
        self.message_type = None
        self.content_length = None
        self.message_content = None

        if message_string is not None:
            self.parse_message(message_string)
        elif message_type is not None and message_content is not None:
            self.message_type = message_type
            self.message_content = message_content
            self.content_length = len(message_content)
        elif message_type is not None:
            self.message_type = message_type
            self.message_content = ""
            self.content_length = 0

    @property
    def readable_message_type(self) -> str:
        if self.message_type in MessageType.__members__.values():
            return f"{self.message_type} {MessageType(self.message_type).name}"
        else:
            return f"{self.message_type} UNKNOWN"

    @property
    def content_expected_for_message_type(self) -> bool:
        return self.message_type < 500

    def parse_growcube_datetime(datetime_str):
        """
        Parse a datetime string in the format "YYYY@MM@DD@HH@mm@SS" into a datetime object.
        This is the format GrowCube uses to send datetimes
        Args:
            datetime_str (str): The datetime string to parse.
        Returns:
            datetime: A datetime object.
        """
        try:
            return datetime.strptime(datetime_str, "%Y@%m@%d@%H@%M@%S")
        except ValueError as e:
            raise ValueError("Invalid datetime string format") from e

    def format_datetime_for_growcube(dt=datetime.now()):
        """
        Format a datetime object as a string in the format "YYYY@MM@DD@HH@mm@SS".
        Args:
            dt (datetime): A datetime object to format.
        Returns:
            str: The formatted datetime string.
        """
        return dt.strftime("%Y@%m@%d@%H@%M@%S")

    def parse_message(self, message_string):
        # Split the message into sections using '#' as the delimiter
        sections = message_string.strip("#").split("#")

        if len(sections) != 3:
            raise ValueError(
                "Invalid message format. It should have exactly 3 sections."
                + message_string
            )

        # Extract and validate message type
        message_type = sections[0]
        if not message_type.startswith("elea") or not message_type[4:].isdigit():
            raise ValueError("Invalid message type format.")
        self.message_type = int(message_type[4:])

        # Extract and validate content length
        content_length = sections[1]
        if not content_length.isdigit():
            raise ValueError("Invalid content length format.")
        self.content_length = int(content_length)

        # Extract and validate message content
        message_content = sections[2]
        if len(message_content) != self.content_length:
            raise ValueError("Content length does not match actual content length.")
        self.message_content = message_content

    def get_fields(self):
        if not self.message_content:
            return []

        # Split the message content into fields using '@' as the delimiter
        fields = self.message_content.split("@")

        return fields

    def get_message(self):
        if self.message_type is None:
            raise ValueError("message_type not specified")

        if self.message_type < 500:
            return (
                f"elea{self.message_type}#{self.content_length}#{self.message_content}#"
            )
        else:
            return f"ele{self.message_type}"

    def __str__(self):
        return f"Message Type: {self.message_type}, Content Length: {self.content_length}, Content: {self.message_content}"
